crates/stateless/src/validation.rs

```diff
@@ -167,38 +168,36 @@ where
         .try_into_recovered()
         .map_err(|err| StatelessValidationError::SignerRecovery(Box::new(err)))?;
 
-    let mut ancestor_headers: Vec<Header> = witness
+    let mut ancestor_headers: Vec<_> = witness
         .headers
         .iter()
-        .map(|serialized_header| {
-            let bytes = serialized_header.as_ref();
-            Header::decode(&mut &bytes[..])
+        .map(|bytes| {
+            let hash = keccak256(bytes);
+            alloy_rlp::decode_exact::<Header>(bytes)
+                .map(|h| SealedHeader::new(h, hash))
                 .map_err(|_| StatelessValidationError::HeaderDeserializationFailed)
         })
```

Comment:
what was the rationale for this change?

Reply:
Using `decode` ignores trailing bytes leading, which can lead to malleability issues. In the worst case, there have even been consensus failures due to clients handling trailing RLP bytes differently. Since this function is intended for creating execution proofs, I think it is always good practice to use `decode_exact`, even if there are no attack vectors in this particular case.